api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,212 @@
+package org.openmrs.module.pihcore.event;
+
+import org.apache.commons.dbutils.handlers.ScalarHandler;
+import org.apache.commons.lang.BooleanUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.openmrs.module.dbevent.Database;
+import org.openmrs.module.dbevent.DatabaseTable;
+import org.openmrs.module.dbevent.DbEvent;
+import org.openmrs.module.dbevent.DbEventSourceConfig;
+import org.openmrs.module.dbevent.EventConsumer;
+import org.openmrs.module.dbevent.Operation;
+import org.openmrs.module.dbevent.Rocks;
+import org.openmrs.module.dbevent.SqlBuilder;
+
+import java.io.File;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Consumes patient-related events
+ * For performance reasons on larger databases, this does not use a Debezium-generated snapshot, but rather
+ * computes a custom snapshot that is then updated as needed from change events.
+ * When used, this should be configured with "snapshot.mode = schema_only"
+ */
+public class PatientUpdateEventConsumer implements EventConsumer {
+
+    private static final Logger log = LogManager.getLogger(PatientUpdateEventConsumer.class);
+
+    private final DbEventSourceConfig config;
+    private final Database database;
+    private final Map<String, String> patientKeys = new LinkedHashMap<>();
+    private Rocks statusDb;
+    private boolean snapshotInitialized = false;
+
+    public PatientUpdateEventConsumer(DbEventSourceConfig config) {
+        this.config = config;
+        this.database = config.getContext().getDatabase();
+        patientKeys.put("person_id", "person");
+        patientKeys.put("person_a", "person");
+        patientKeys.put("person_b", "person");
+        patientKeys.put("order_id", "orders");
+        patientKeys.put("patient_program_id", "patient_program");
+        patientKeys.put("encounter_id", "encounter");
+        patientKeys.put("visit_id", "visit");
+        patientKeys.put("allergy_id", "allergy");
+        patientKeys.put("appointment_id", "appointmentscheduling_appointment");
```

Comment:
Do we need obs here?

Reply:
No.  I may rework the approach here.  This basically represents the possible joins based on key found in the event.   eg. if a table has "order_id", then it needs to join on "orders".  Obs has a person_id, so it's covered by the first entry.